src/Polly.Core/Timeout/TimeoutResilienceStrategy.cs

```diff
@@ -0,0 +1,108 @@
+        {
+            // do nothing
+            return await callback(context, state).ConfigureAwait(context.ContinueOnCapturedContext);
+        }
+
+        var previousToken = context.CancellationToken;
+        using var cancellationSource = new CancellationTokenSource();
+        _timeProvider.CancelAfter(cancellationSource, timeout);
+        context.CancellationToken = cancellationSource.Token;
+
+        CancellationTokenRegistration? registration = null;
+
+        if (previousToken.CanBeCanceled)
+        {
+            registration = previousToken.Register(static state => ((CancellationTokenSource)state!).Cancel(), cancellationSource, useSynchronizationContext: false);
+        }
+
+        try
+        {
+            if (_strategy == TimeoutStrategy.Pessimistic)
+            {
+                callback = MakeCancellable(callback);
+            }
+
+            return await callback(context, state).ConfigureAwait(context.ContinueOnCapturedContext);
+        }
+        catch (OperationCanceledException e) when (cancellationSource.IsCancellationRequested && !previousToken.IsCancellationRequested)
+        {
+            _telemetry.Report(TimeoutConstants.OnTimeoutEvent, context);
+
+            await _onTimeout(new OnTimeoutArguments(context, e, timeout)).ConfigureAwait(context.ContinueOnCapturedContext);
+
+            throw new TimeoutRejectedException("The timeout resilience strategy did not complete within the timeout.", e);
+        }
+        finally
+        {
+            context.CancellationToken = previousToken;
+
+            if (registration.HasValue)
+            {
+#if NETCOREAPP
+                await registration.Value.DisposeAsync().ConfigureAwait(context.ContinueOnCapturedContext);
+#else
+                registration.Value.Dispose();
+#endif
+            }
+        }
+    }
+
+    private static Func<ResilienceContext, TState, ValueTask<TResult>> MakeCancellable<TState, TResult>(Func<ResilienceContext, TState, ValueTask<TResult>> callback)
+    {
+        return async (context, state) =>
+        {
+            var timeoutTask = AsTask<TResult>(context.CancellationToken);
+            var actionTask = callback(context, state).AsTask();
+
+            var completedTask = await Task.WhenAny(actionTask, timeoutTask).ConfigureAwait(context.ContinueOnCapturedContext);
+
+            return await completedTask.ConfigureAwait(context.ContinueOnCapturedContext);
+        };
+    }
+
+    private static Task<TResult> AsTask<TResult>(CancellationToken cancellationToken)
+    {
+        var tcs = new TaskCompletionSource<TResult>();
+
+        // A generalized version of this method would include a hot-path returning a canceled task (rather than setting up a registration)
+        // if (cancellationToken.IsCancellationRequested) on entry.  This is omitted, since we only start the timeout countdown in the token _after calling this method.
```

Comment:
```suggestion
        // if (cancellationToken.IsCancellationRequested) on entry.  This is omitted, since we only start the timeout countdown in the token after calling this method.
```